File: skiplists/skiplists.swift
//
//  skiplists.swift
//  skiplists
//
//  Created by Peter da Silva on 5/25/16.
//  Copyright © 2016 Flightaware. All rights reserved.
//

import Foundation

let randomProbability = 0.5

func SkipListRandomLevel(maxLevel: Int) -> Int {
    var newLevel = 1
    while drand48() < randomProbability && newLevel < maxLevel {
        newLevel += 1
    }
    return newLevel
}

func SkipListMaxLevel(maxNodes: Int) -> Int {
        let logMaxNodes = log(Double(maxNodes)) / log(1.0 / randomProbability)
        return Int(round(logMaxNodes))
}

class SLNode<Key: Comparable, Value: Equatable> {
    let key: Key?
    var values: [Value]
    var level: Int
    var next: [SLNode<Key, Value>?]
    init(_ key: Key?, value: Value? = nil, maxLevel: Int, level: Int = 0) {
        self.key = key
        if let v = value {
            self.values = [v]
        } else {
            self.values = []
        }
        self.level = (level > 0) ? level : SkipListRandomLevel(maxLevel)
        self.next = Array<SLNode<Key, Value>?>(count: maxLevel, repeatedValue: nil)
    }
}

public class SkipList<Key: Comparable, Value: Equatable>: SequenceType {
    let head: SLNode<Key, Value>
    var maxLevel: Int
    var level: Int
    
    public init(maxLevel: Int) {
        self.maxLevel = maxLevel
        self.level = 1
        self.head = SLNode<Key, Value>(nil, maxLevel: maxLevel, level: maxLevel)
    }
    
    public convenience init(maxNodes: Int) {
        self.init(maxLevel: SkipListMaxLevel(maxNodes))
    }
    
    func search(greaterThanOrEqualTo key: Key) -> SLNode<Key, Value>? {
        var x = head
        
        // look for the key
        for i in (1 ... self.level).reverse() {
            while let next = x.next[i-1] where next.key < key {
                x = next
            }
        }
        
        return x.next[0]
    }
    
    public func search(greaterThanOrEqualTo key: Key) -> [Value] {
        if let array = search(greaterThanOrEqualTo: key)?.values {
            return array
        } else {
            return []
        }
    }
    
    func search(equalTo key: Key) -> SLNode<Key, Value>? {
        // Check for an exact match
        if let x: SLNode<Key, Value> = search(greaterThanOrEqualTo: key) where x.key == key {
            return x
        } else {
            return nil
        }
    }
    
    public func exists(key: Key) -> Bool {
        let x: SLNode<Key, Value>? = search(equalTo: key)
        return x != nil
    }
    
    public func exists(key: Key, value: Value) -> Bool {
        if let array = search(equalTo: key)?.values {
            return array.contains(value)
        }
        return false;
    }
    
    public func search(equalTo key: Key) -> [Value] {
        if let array = search(equalTo: key)?.values {
            return array
        } else {
            return []
        }
    }
    
    // Replace an entry in a skiplist - for optional keys
    public func replace(newKey: Key?, inout keyStore: Key?, value: Value) {
        // no change - no work
        if newKey == keyStore {
            return
        }
        
        // showtime -- remove the old entry, update the keystore, insert the new value
        if let k = keyStore {
            delete(k, value: value)
        }
        keyStore = newKey
        if let k = newKey {
            insert(k, value: value)
        }
    }
    
    // Replace an entry in a skiplist - for non-optional keys
    public func replace(newKey: Key, inout keyStore: Key, value: Value) {
        // no change - no work
        if newKey == keyStore {
            return
        }
        
        // showtime -- remove the old entry, update the keystore, insert the new value
        delete(keyStore, value: value)
        keyStore = newKey
        insert(keyStore, value: value)
    }
        
    public func insert(key: Key, value newValue: Value) {
        var update = Array<SLNode<Key, Value>?>(count: maxLevel, repeatedValue: nil)
        var x = head
        var i: Int
        
        // look for the key, and save the previous nodes all the way down in the update[] list
        i = self.level
        while i >= 1 {
            while let next = x.next[i-1] where next.key < key {
                x = next
            }
            update[i-1] = x
            i -= 1
        }
        
        // If we haven't run off the end...
        if let next = x.next[0]  {
            x = next
            
            // If we're looking at the right key already, then there's nothing to insert. Just add
            // the new value to the values array.
            if x.key == key {
                if x.values.contains(newValue) {
                    return
                }
                
                x.values += [newValue]
                
                return
            }
        }
        
        // Pick a random level for the new node
        let level = SkipListRandomLevel(maxLevel)
        
        // If the new node is higher than the current level, fill up the update[] list
        // with head
        while level > self.level {
            self.level += 1
            update[self.level-1] = self.head
        }
        
        // make a new node and patch it in to the saved nodes in the update[] list
        let newNode = SLNode<Key, Value>(key, value: newValue, maxLevel: maxLevel, level: level)
        i = 1
        while i <= level {
            newNode.next[i-1] = update[i-1]!.next[i-1]
            update[i-1]!.next[i-1] = newNode
            i += 1
        }
    }
    
    public func delete(key: Key, value: Value) -> Bool {
        var update = Array<SLNode<Key, Value>?>(count: maxLevel, repeatedValue: nil)
        var x = head
        var i: Int
        
        // look for the key, and save the previous nodes all the way down in the update[] list
        i = self.level
        while i >= 1 {
            while let next = x.next[i-1] where next.key < key {
                x = next
            }
            update[i-1] = x
            i -= 1
        }
        
        // check if run off end of list, nothing to do,
        guard let next = x.next[0] else {
            return false
        }
        
        // Point to the node we're maybe going to delete, if it matches
        x = next
        
        // Look for a key match
        if x.key != key {
            return false
        }
        
        // look for match in values
        guard let foundIndex = x.values.indexOf(value) else {
            // If we didn't find a matching value, we didn't actually find a match
            return false
        }
        
        // Remove the value we found, and if it wasn't the last one return success
        x.values.removeAtIndex(foundIndex)
        if(x.values.count > 0) {
            return true
        }

        // Now we've found a value, deleted it, and emptied the values list, we can delete this whole node

        // point all the previous node to the new next node
        i = 1
        while i <= self.level {
            if update[i-1]!.next[i-1] != nil && update[i-1]!.next[i-1] !== x {
                break
            }
            update[i-1]!.next[i-1] = x.next[i-1]
            i += 1
        }
            
        // if that was the biggest node, and we can see the end of the list from the head,
        // lower the list until we're pointing at a node
        while self.level > 1 && self.head.next[self.level-1] == nil {
            self.level -= 1
        }
        
        return true
    }
    
    public func generate() -> AnyGenerator<(Key, Value)> {
        var row = head
        var index = -1
        
        return AnyGenerator<(Key, Value)> {
            if index < 0 || index >= row.values.count {
                repeat {
                    guard let next = row.next[0] else { return nil }
                    row = next
                } while row.values.count == 0
                index = 0
            }
            let next = row.values[index]
            index += 1
            return (row.key!, next)
        }
    }
    
    func query(from start: Key?, through end: Key?) -> Query<Key, Value> {
        return query(min: start, max: end, minEqual: true, maxEqual: true)
    }
    
    func query(from start: Key?, to end: Key?) -> Query<Key, Value> {
        return query(min: start, max: end, minEqual: true, maxEqual: false)
    }

    func query(greaterThanOrEqual key: Key?) -> Query<Key, Value> {
        return query(min: key, minEqual: true)
    }
    
    func query(greaterThan key: Key?) -> Query<Key, Value> {
        return query(min: key, minEqual: false)
    }

    func query(lessThanOrEqual key: Key?) -> Query<Key, Value> {
        return query(max: key, maxEqual: true)
    }
    
    func query(lessThan key: Key?) -> Query<Key, Value> {
        return query(max: key, maxEqual: false)
    }
    
    func query(min min: Key? = nil, max: Key? = nil, minEqual: Bool = false, maxEqual: Bool = false) -> Query<Key, Value> {
        return Query<Key, Value>(list: self, min: min, max: max, minEqual: minEqual, maxEqual: maxEqual)
    }
}

File: skiplists/skiplists.c
//
//  skiplists.c
//  skiplists
//
//  Created by Peter da Silva on 6/11/16.
//  Copyright © 2016 Flightaware. All rights reserved.
//

#include "skiplists.h"

// DEBUGGER
void dumpNode(struct C_SkipListNode *node, char *name)
{
    if(name) fprintf(stderr, "name = %s ", name);
    if(!node) {
        fprintf(stderr, "NULL");
        return;
    }
    fprintf(stderr, "%lx", (long)node);
    if(node->keyString) fprintf(stderr, " '%s'", node->keyString);
    fprintf(stderr, " level=%d", node->level);
}

void dumpList(struct C_SkipList *list, char *name)
{
    if(name) fprintf(stderr, "%s: ", name);
    if(!list) {
        fprintf(stderr, "NULL");
        return;
    }
    dumpNode(list->head, "[head");
    fprintf(stderr, "]; maxLevels = %d; level = %d; type = %d;\n", list->maxLevels, list->level, list->type);
}

void dumpSearch(struct C_SkipListSearch *search, char *name)
{
    int i;
    if(name) fprintf(stderr, "SEARCH %s:\n", name);
    dumpList(search->parent, "parent");
    for(i = 0; i < search->parent->maxLevels; i++) {
        fprintf(stderr, "update[%d] = ", i);
        dumpNode(search->update[i], NULL);
        fprintf(stderr, ";\n");
    }
}


double randomProbability = 0.5;

int randomLevel(int maxLevels)
{
    int newLevel = 1;
    while(drand48() < randomProbability && newLevel < maxLevels-1) {
        newLevel += 1;
    }
    return newLevel;
}

// maxlevels is the max depth of this skiplist
struct C_SkipListNode *newSkipListNode(int maxLevels, int level)
{
    struct C_SkipListNode *node = malloc(sizeof *node);
    node->keyString = NULL;
    node->value = NULL;
    node -> level = level;
    node->next = malloc(maxLevels * sizeof node);

    int i;
    for(i = 0; i < maxLevels; i++)
        node->next[i] = NULL;

    return node;
}

void destroySkipListNode(struct C_SkipListNode *node)
{
    if(node) {
        if(node -> keyString) free(node->keyString);
        free(node->next);
        free(node);
    }
}

struct C_SkipList *newSkipList(int maxLevels, int type)
{
    struct C_SkipList *list = malloc(sizeof *list);
    list->head = newSkipListNode(maxLevels, 0);
    list->maxLevels = maxLevels;
    list->level = 1;
    list->type = type;
    return list;
}

void destroySkipList(struct C_SkipList *list)
{
    if(list) {
        while(list->head) {
            struct C_SkipListNode *next = list->head->next[0];
            destroySkipListNode(list->head);
            list->head = next;
        }
        free(list);
    }
}


struct C_SkipListSearch *newSkipListSearch(struct C_SkipList *parent)
{
    struct C_SkipListSearch *search = malloc(sizeof *search);
    if(!search) return NULL;
    search->update = malloc(parent->maxLevels * sizeof (struct C_SkipListNode));
    if(!search->update) { free(search); return NULL; }
    search->parent = parent;
    search->node = NULL;
    search->state = SEARCH_STATE_NONE;
    
    int i;
    for(i = 0; i < parent->maxLevels; i++)
        search->update[i] = NULL;

    return search;
}

void destroySkipListSearch(struct C_SkipListSearch *search)
{
    if(search) {
        free(search->update);
        free(search);
    }
}

// Shortcut search set up to traverse the whole list
void traverseSkipList(struct C_SkipListSearch *search)
{
    if(search->parent->head) {
        search->node = search->parent->head->next[0];
    } else {
        search->node = NULL;
    }
    search->state = search->node ? SEARCH_STATE_TRAVERSE : SEARCH_STATE_NONE;
}

int searchSkipListString(struct C_SkipListSearch *search, const char *keyString)
{
    struct C_SkipList *list = search->parent;
    struct C_SkipListNode *x = list->head;
    int i;
    
    // dumpSearch(search, "before search");
    for(i = list->level; i >= 1; i--) {
        while(x->next[i-1] != NULL && strcmp(x->next[i-1]->keyString, keyString) < 0) {
            x = x->next[i-1];
        }
        search->update[i-1] = x;
    }
    // dumpSearch(search, "after search");
    // dumpNode(x, "x"); fprintf(stderr, ";\n");
    if (x->next[i-1] == NULL) {
        search->state = SEARCH_STATE_NOT_FOUND;
    } else {
        search->state = SEARCH_STATE_FOUND;
        search->node = search->update[0];
    }
    return search->state = SEARCH_STATE_FOUND;
}

int searchMatchedExactString(struct C_SkipListSearch *search, const char *keyString)
{
    if(search->state != SEARCH_STATE_FOUND) return 0;
    if(search->node == NULL) return 0;
    if(search->node->keyString == NULL) return keyString == NULL;

    return strcmp(search->node->keyString, keyString) == 0;
}

const char *getMatchedKeyString(struct C_SkipListSearch *search)
{
    if(search->state != SEARCH_STATE_FOUND && search->state != SEARCH_STATE_TRAVERSE) return NULL;
    if(search->node == NULL) return NULL;
    
    return search->node->keyString;
}

void *getMatchedValue(struct C_SkipListSearch *search)
{
    if(search->state != SEARCH_STATE_FOUND && search->state != SEARCH_STATE_TRAVERSE) return NULL;
    if(search->node == NULL) return NULL;

    return search->node->value;
}

int setMatchedValue(struct C_SkipListSearch *search, void *value)
{
    if(search->state != SEARCH_STATE_FOUND) return 0;
    if(search->node == NULL) return 0;
    
    search->node->value = value;
    return 1;
}

int advanceSearchNode(struct C_SkipListSearch *search)
{
    if(search->state != SEARCH_STATE_FOUND && search->state != SEARCH_STATE_TRAVERSE) return 0;
    if(search->node == NULL) return 0;
    search->node = search->node->next[0];
    if(search->node == NULL) {
        search->state = SEARCH_STATE_NONE;
        return 0;
    } else {
        search->state = SEARCH_STATE_TRAVERSE;
        return 1;
    }
}

// Generic insert for all types
int insertBeforePossibleMatch(struct C_SkipListSearch *search, struct C_SkipListNode *newNode, int level)
{
    struct C_SkipList *list = search->parent;
    if(search->state != SEARCH_STATE_FOUND && search->state != SEARCH_STATE_NOT_FOUND) return 0;
    search->state = SEARCH_STATE_NONE;
    search->node = NULL;
    
    // dumpSearch(search, "insert");
    // If the new node is higher than the current level, fill up the update[] list
    // with head
    while(level >= list->level) {
        search->update[list->level-1] = list->head;
        list->level += 1;
    }
    
    // patch new node in to the saved nodes in the update[] list
    int i;
    for(i = 1; i <= level; i ++) {
        newNode->next[i-1] = search->update[i-1]->next[i-1];
        search->update[i-1]->next[i-1] = newNode;
    }
    return 1;
}

int searchCanInsert(struct C_SkipListSearch *search)
{
    return search->state == SEARCH_STATE_FOUND || search->state == SEARCH_STATE_NOT_FOUND;
}

// Specific insert for string
int insertBeforePossibleMatchString(struct C_SkipListSearch *search, const char *keyString, void *value)
{
    struct C_SkipList *list = search->parent;
    if(search->state != SEARCH_STATE_FOUND && search->state != SEARCH_STATE_NOT_FOUND) return 0;

    // Pick a random level for the new node
    int level = randomLevel(list->maxLevels);

    // Create the new node
    struct C_SkipListNode *newNode = newSkipListNode(list->maxLevels, level);
    strcpy(newNode->keyString = malloc(strlen(keyString) + 1), keyString);
    newNode -> value = value;
    
    // Call the general routine for the tricky stuff
    if (!insertBeforePossibleMatch(search, newNode, level)) { // can't happen
        destroySkipListNode(newNode);
        return 0;
    }
    return 1;
}

int deleteMatchedNode(struct C_SkipListSearch *search)
{
    struct C_SkipList *list = search->parent;
    if(search->state != SEARCH_STATE_FOUND) return 0;
    search->state = SEARCH_STATE_NONE;
    search->node = NULL;
    
    struct C_SkipListNode *x = search->update[0];
    
    // point all the previous node to the new next node
    int i;
    for(i = 1; i < list->level; i ++) {
        if(search->update[i-1]->next[i-1] != x)
            break;
        search->update[i-1]->next[i-1] = x->next[i-1];
    }
    
    // if that was the biggest node, and we can see the end of the list from the head,
    // lower the list until we're pointing at a node
    while(list->level > 1 && list->head->next[list->level-1] == NULL) {
        list->level--;
    }
    
    // Dispose of the node, because we're doing memory management
    destroySkipListNode(x);
    
    return 1;
}